pkg/segment/set.go

```diff
@@ -1243,6 +1469,96 @@ func (gap GapIterator) NextGap() GapIterator {
 	return seg.NextGap()
 }
 
+// NextLargeEnoughGap returns the iterated gap's first next gap with larger length than minSize.
+// If not found, return a terminal gap iterator (does NOT include this gap itself).
+func (gap GapIterator) NextLargeEnoughGap(minSize Key) GapIterator {
```

Comment:
This package tries to avoid recursion (Go stack handling is somewhat expensive, and Go doesn't implement TCO). I also find the logic here a bit hard to follow. How about:

```
func (gap GapIterator) NextSizedGap(min Key) GapIterator {
  seg :=  gap.NextSegment()
  if !seg.Ok() {
    return GapIterator{}
  }
  return seg.NextSizedGap(min)
}

func (seg Iterator) NextSizedGap(min Key) GapIterator {
  return firstSizedGapAfter(seg.node, seg.index, min)
}

func firstSizedGapAfter(n *node, index int, min Key) GapIterator {
	// Walk up the tree until we find a node for which node.maxGap >= min.
	for n.maxGap < min {
		if n.parent == nil {
			return GapIterator{}
		}
		n, index = n.parent, n.parentIndex
	}
	// If n is a leaf, check all gaps after index.
	if !n.hasChildren {
		for i := index + 1; i <= n.nrSegments; i++ {
			if gap := (GapIterator{n, i}); gap.Range().Length() >= min {
				return gap
			}
		}
		if n.parent == nil {
			return GapIterator{}
		}
		n, index = n.parent, n.parentIndex
	}
	// n and its ancestors are not leaves; check all subtrees after index.
	for {
		for i := index + 1; i <= n.nrSegments; i++ {
			if n.children[i].maxGap >= min {
				return firstSizedGapIn(n.children[i], min)
			}
		}
		if n.parent == nil {
			return GapIterator{}
		}
		n, index = n.parent, n.parentIndex
	}
}

// firstSizedGapIn returns the first gap in the tree rooted by n whose length
// is at least min.
//
// Preconditions: n.maxGap >= min.
func firstSizedGapIn(n *node, min Key) GapIterator {
outer:
	for n.hasChildren {
		for i := 0; i <= n.nrSegments; i++ {
			if n.children[i].maxGap >= min {
				n = n.children[i]
				continue outer
			}
		}
		panic(fmt.Sprintf("node.maxGap == %d, but contains no subtrees with maxGap >= %d: %v", node.maxGap, min, node))
	}
	for i := 0; i < n.nrSegments; i++ {
		if gap := (GapIterator{n, i}); gap.Range().Length() >= min {
			return gap
		}
	}
	panic(fmt.Sprintf("node.maxGap == %d, but contains no gaps with length >= %d: %v", node.maxGap, min, node))
}
```

Reply:
To be honest, I'm relunctant to modify the algorithm at such large scale unless there's any bug or error spotted. I agree to avoid recursion if it's the case as you described so I follow your suggestions above for the maintainence of maxGap part because they are always called. But for the search part, the overall efficiency is already a lot better than the original linear scan and this is generally "single-time" cost, I feel it's somehow unnecessary.

Reply:
The code is sufficiently unclear that spotting bugs and errors is difficult. We care about performance.

Reply:
It's clear to me and I can help explain here. Let's take NextlargeEnoughGap as example. 
The idea very straightforward, if the calling gap is living in a node with maxGap larger than requirement and it's not the trailing gap on a leaf node (i.e. no next gap in this node), we iterates the following gaps and check if the gap length satisfies accordingly. If not found, which means the maxGap within the node is before or is the calling gap itself, go to the parent and recurse. Otherwise, crawl up the tree.
Since we iterate towards next, the case of trailing gap in internal node is equivalently translated to be the first gap in the subtree within this gap. Actually this case could never be met in real use case because FindNextLargeEnoughGap is only called by LowerBoundGap's returned gap which is ensured to be on leaf node.
This algorithm is shown to be correct through many test cases on corner cases and huge number of pure random tests. 
I definitely care about performance and that's the whole reason I'm working on this project. With this improvement, it's already much much better than the original linear scan thus performance is already significantly improved. I don't think we need to care about the small overhead of stack handling by recursion at the cost of code elegancy and introducing much more complexity.

Reply:
If there's anything still unclear or not comfortable to you, feel free to let me know. I'm open to suggestion.